src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -1195,17 +1463,735 @@ private static void ReportReadOnlyError(Symbol symbol, SyntaxNode node, BindValu
+                    }
+
+                    var eventSymbol = eventAccess.EventSymbol;
+
+                    // field-like events that are static or belong to reference types can ref escape anywhere
+                    if (eventSymbol.IsStatic || eventSymbol.ContainingType.IsReferenceType)
+                    {
+                        return Binder.ExternalScope;
+                    }
+
+                    // for other events defer to the receiver.
+                    return GetRefEscape(eventAccess.ReceiverOpt, scopeOfTheContainingExpression);
+
+                case BoundKind.Call:
+                    var call = (BoundCall)expr;
+
+                    var methodSymbol = call.Method;
+                    if (methodSymbol.RefKind == RefKind.None)
+                    {
+                        break;
+                    }
+
+                    return GetInvocationEscapeScope(
+                        call.ReceiverOpt,
+                        methodSymbol.Parameters,
+                        call.Arguments,
+                        call.ArgumentRefKindsOpt,
+                        call.ArgsToParamsOpt,
+                        scopeOfTheContainingExpression);
+
+                case BoundKind.IndexerAccess:
+                    var indexerAccess = (BoundIndexerAccess)expr;
+                    var indexerSymbol = indexerAccess.Indexer;
+
+                    return GetInvocationEscapeScope(
+                        indexerAccess.ReceiverOpt,
+                        indexerSymbol.Parameters,
+                        indexerAccess.Arguments,
+                        indexerAccess.ArgumentRefKindsOpt,
+                        indexerAccess.ArgsToParamsOpt,
+                        scopeOfTheContainingExpression);
+
+                case BoundKind.PropertyAccess:
+                    var propertyAccess = (BoundPropertyAccess)expr;
+
+                    // not passing any arguments/parameters
+                    return GetInvocationEscapeScope(
+                        propertyAccess.ReceiverOpt,
+                        default,
+                        default,
+                        default,
+                        default,
+                        scopeOfTheContainingExpression);
+            }
+
+            // At this point we should have covered all the possible cases for anything that is not a strict RValue.
+            return scopeOfTheContainingExpression;
+        }
+
+        /// <summary>
+        /// A counterpart to the GetRefEscape, which validates if given escape demand can be met by the expression.
+        /// The result indicates whether the escape is possible. 
+        /// Additionally, the method emits diagnostics (possibly more than one, recursively) that would help identify the cause for the failure.
+        /// </summary>
+        internal static bool CheckRefEscape(SyntaxNode node, BoundExpression expr, uint escapeFrom, uint escapeTo, bool checkingReceiver, DiagnosticBag diagnostics)
```

Comment:
I saw your comment in description about the split between "Get" and "Check" methods. I wonder if the expected perf gain really warrants the additional code complexity/duplication. Could a single set of methods achieve the purpose?

Reply:
Not just a perf gain. It is easier to concentrate on Get and Check methods separately. 
If I had to make the same method that work as both, it could be harder to do correctly. Perhaps, once we are done, we can refactor this into one method. For now I feel it would just make the uber method very complex.

---
In reply to: [135401583](https://github.com/dotnet/roslyn/pull/21334#discussion_r135401583) [](ancestors = 135401583)